src/hotspot/share/gc/shenandoah/shenandoahControlThread.hpp

```diff
@@ -122,11 +122,10 @@ class ShenandoahControlThread: public ConcurrentGCThread {
   ~ShenandoahControlThread();
 
   // Handle allocation failure from normal allocation.
-  // Blocks until memory is available.
-  void handle_alloc_failure(ShenandoahAllocRequest& req);
+  // Optionally blocks while collector is handling the failure.
```

Comment:
Who is allowed to call the non-blocking version? What is the semantics of block or don't block? Can the documentation be extended to elaborate on its use?

Reply:
When a mutator allocation fails, it will block on this call and wait to be notified by the control thread to retry the allocation. The notification follows a degenerated or full GC.

Reply:
With this change, the mutator is also allowed to call the non-blocking version when it cannot allocate because the GC overhead has been exceeded. I didn't make sense to me to have the mutator block and wait for a GC cycle when the GC threshold is already exceeded.

Reply:
OK, thanks, that makes sense.